drivers/wifi/esp8266/esp8266.c

```diff
@@ -0,0 +1,784 @@
+	memcpy(&cmd[len], params->ssid, params->ssid_length);
+	len += params->ssid_length;
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+		memcpy(&cmd[len], params->psk, params->psk_length);
+		len += params->psk_length;
+		ecn = 3;
+	} else {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+	}
+
+	snprintk(&cmd[len], sizeof(cmd) - len, "\",%d,%d", params->channel,
+		 ecn);
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, cmd, &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static int esp8266_mgmt_ap_disable(struct device *dev)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+CWMODE_CUR=1", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static void esp8266_init_work(struct k_work *work)
+{
+	struct esp8266_data *dev;
+	int ret;
+	static struct setup_cmd setup_cmds[] = {
+		/* turn off echo */
+		SETUP_CMD_NOHANDLE("ATE0"),
+		/* enable multiple socket support */
+		SETUP_CMD_NOHANDLE("AT+CIPMUX=1"),
+		SETUP_CMD_NOHANDLE("AT+CWMODE_CUR=1"),
+		/* only need ecn,ssid,rssi,channel */
+		SETUP_CMD_NOHANDLE("AT+CWLAPOPT=0,23"),
+#if defined(CONFIG_WIFI_ESP8266_PASSIVE_TCP)
+		SETUP_CMD_NOHANDLE("AT+CIPRECVMODE=1"),
+#endif
+		SETUP_CMD("AT+CIPAPMAC_CUR?", "+CIPAPMAC_CUR:",
+			  on_cmd_cipapmac, 1U, ""),
+	};
+
+	dev = CONTAINER_OF(work, struct esp8266_data, init_work);
+
+	ret = modem_cmd_handler_setup_cmds(&dev->mctx.iface,
+					   &dev->mctx.cmd_handler, setup_cmds,
+					   ARRAY_SIZE(setup_cmds),
+					   &dev->sem_response,
+					   ESP_INIT_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("Init failed %d", ret);
+		return;
+	}
+
+	net_if_set_link_addr(dev->net_iface, dev->mac_addr,
+			     sizeof(dev->mac_addr), NET_LINK_ETHERNET);
+
+	LOG_INF("ESP8266 Wi-Fi ready");
+
+	net_if_up(dev->net_iface);
+
+	k_sem_give(&dev->sem_if_up);
+}
+
+static void esp8266_reset(struct esp8266_data *dev)
+{
+	int ret;
+
+	if (net_if_is_up(dev->net_iface)) {
+		net_if_down(dev->net_iface);
+	}
+
+#if defined(DT_INST_0_ESPRESSIF_ESP8266_WIFI_RESET_GPIOS_PIN)
+	modem_pin_write(&dev->mctx, WIFI_RESET, 0);
+	k_sleep(K_MSEC(100));
+	modem_pin_write(&dev->mctx, WIFI_RESET, 1);
+#else
+	ret = modem_cmd_send(&dev->mctx.iface, &dev->mctx.cmd_handler,
+			     NULL, 0, "AT+RST", &dev->sem_response,
+			     ESP_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("Failed to reset device: %d", ret);
+		return;
+	}
+#endif
+
+	LOG_INF("Waiting for interface to come up");
+
+	ret = k_sem_take(&dev->sem_if_up, ESP_INIT_TIMEOUT);
+	if (ret == -EAGAIN) {
+		LOG_ERR("Timeout waiting for interface");
+	}
+}
+
+static void esp8266_iface_init(struct net_if *iface)
+{
+	struct device *dev = net_if_get_device(iface);
+	struct esp8266_data *data = dev->driver_data;
+
+	atomic_clear_bit(iface->if_dev->flags, NET_IF_UP);
```

Comment:
If you do not want the network interface to be up when device is initialized,
just call `net_if_flag_set(iface, NET_IF_NO_AUTO_START);` here.